### PR TITLE
[librustdoc] Only split lang string on `,`, ` `, and `\t`

### LINKED ISSUE #78344: Rustdoc doctest attribute splitting is probably too liberal
I was thinking about [proposing a new doctest attribute](https://github.com/casey/rfcs/blob/doctest-name/text/0000-doctest-name.md) of the form `name=TEXT` that would allow giving a documentation tests names which would be printed when they run. 

I was looking at what it would take to implement this change, so I checked out the way that doctest attributes are parsed from markdown info strings, and found that they are split using the following code:

```rust
let tokens = string.split(|c: char| !(c == '_' || c == '-' || c.is_alphanumeric()));
```

("info string" is the name that common mark and GFM use for the text that follows a ```` ``` ````.)

This conflicts with the `name=foo` syntax that I was hoping to use, since the attribute would be split on the `=`.

I think this behavior also conflicts with how people think the feature works, since a few people told me that attributes are split on `,`, which is also [what Clippy does](https://github.com/rust-lang/rust-clippy/blob/8b70b84639a368eafbc278dbdd93a6286abf23b2/clippy_lints/src/doc.rs#L409).

My feeling is that the current attribute splitting is probably too liberal in what it accepts, and prevents otherwise desirable attributes, such as attributes of the form `foo=bar`, or attributes with more freeform text, from being possible.

A few thoughts:

- Is this worth worrying about at all? It's unfortunate, but it's not a *huge* deal, since workarounds like having attributes of the form `foo-bar` are possible.

- If nobody (according to crater) is relying on the liberal splitting, would it be acceptable to consider this to be a bug, and change it to just split on `,`? Of course, this could break non-public code.

- Would it be worth doing as part of an edition? It would be pretty easy to automatically transform info strings from being whatever-split to being comma split as part of an edition upgrade. It would also be easy to create a lint or warning for it, just check that the current splitting rules and comma splitting produce the same attributes, and warn if they don't.

### PR DESCRIPTION
**Pull request by casey**
Split markdown lang strings into tokens on `,`.

The previous behavior was to split lang strings into tokens on any
character that wasn't a `_`, `_`, or alphanumeric.

This is a potentially breaking change, so please scrutinize! See discussion in #78344.

I noticed some test cases that made me wonder if there might have been some reason for the original behavior:

```
t("{.no_run .example}", false, true, Ignore::None, true, false, false, false, v(), None);
t("{.sh .should_panic}", true, false, Ignore::None, false, false, false, false, v(), None);
t("{.example .rust}", false, false, Ignore::None, true, false, false, false, v(), None);
t("{.test_harness .rust}", false, false, Ignore::None, true, true, false, false, v(), None);
```

It seemed pretty peculiar to specifically test lang strings in braces, with all the tokens prefixed by `.`.

I did some digging, and it looks like the test cases were added way back in [this commit from 2014](https://github.com/rust-lang/rust/commit/3fef7a74ca9a) by @skade.

It looks like they were added just to make sure that the splitting was permissive, and aren't testing that those strings in particular are accepted.

Closes https://github.com/rust-lang/rust/issues/78344.

### CODE REVIEW

**Comment by jyn514** [comment]
@rfcbot fcp merge

This changes the parsing of language strings for code blocks so that only `,` is accepted as a delimiter. Previously, any character that couldn't be part of an identifier was treated as a delimiter. This gives rustdoc increased flexibility to distinguish betweeen delimiters, such as for https://github.com/casey/rfcs/blob/doctest-name/text/0000-doctest-name.md.

This is a breaking change, but in practice it seems unlikely anyone is using this. It's possible that we could find out with a crater run if crater allows denying specific warnings.

Possible alternatives:
- Only disallow `=`. This still unblocks the extension @casey proposes in the linked RFC, but allows using all other non-identifiers as delimiters. Any future extensions to lang strings will have to reserve more delimiters at that time.
- Keep the status quo. This is not a breaking change, but means that rustdoc will never be able to add more extensions to language strings.

**Comment by jyn514** [comment]
@casey as-is, this breaks trailing text after spaces:

```
/// ignore (tidy)
/// goo();
/// ```
pub fn bar() -> usize { 2 }
```
```
6: @has check failed
	`XPATH PATTERN` did not match
	// @has foo/fn.bar.html '//*[@class="tooltip ignore"]/span' "This example is not tested"
```
What do you think about changing this to split on either spaces or `,` instead?

**Comment by casey** [comment]
@jyn514 Sounds good to me! I think using spaces is common, particularly with GitHub flavored markdown, so splitting on spaces seems reasonable. I also split on tabs, since it seems weird to split on spaces but not tabs, and since users sometimes use tabs for alignment.

**Comment by GuillaumeGomez** [comment]
Hum also: we use such headers in rust tests:

``````
```rust,ignore (whatever is the reason)
``````

So splitting on whistepaces and tabs might not be a good idea...

**Comment by casey** [comment]
Unrecognized tokens are ignored, so this shouldn't change the set of recognized tokens in the above example.

Before the change, `rust,ignore (whatever is the reason)` will be split into `rust`, `ignore`, `whatever`, `is`, `the`, `reason`, and all but `rust` and `ignore` are ignored.

After the change, `rust,ignore (whatever is the reason)` will be split into `rust`, `ignore`, `(whatever`, `is`, `the`, `reason)`, and all but `rust` and `ignore` are ignored.

**Comment by GuillaumeGomez** [comment]
Yes but what's the point of splitting on whitespaces and tabs?

**Comment by casey** [comment]
If it didn't split on spaces, then something like `ignore (tidy)` will be split as `ignore (tidy)` (one token) and so won't work. This seems to be somewhat common, so it seems better not to break it.

**Comment by GuillaumeGomez** [comment]
Oh I see, good point!

**Comment by nikomatsakis** [comment]
Nominating for @rust-lang/lang meeting to draw attention to the PR.

**Comment by nikomatsakis** [comment]
I'd like to see crater results, but I think this is reasonable. It'd be nice to have a survey of how other markdown systems approach this though.

**Comment by Manishearth** [comment]
Also would like to see crater results, but I'm fine with this otherwise

@rfcbot approve

**Comment by casey** [comment]
I did a mini survey of other markdown systems.

- **Github**: Splits on `,`, ` `, and `\t`.
- **Gitlab**: Doesn't split on anything. Entire info string is used verbatim, so `rust,foo` and `rust foo` disable syntax highlighting.
- **Misune**, a popular python markdown library: Splits on `,`, all `,` separated items are added to the class list in the generated HTML. Everything after the first space or tab is ignored, so `rust foo` has `class=language-rust`, and ignores `foo`.
- **Remark**, a popular JS markdown library: Splits on `,` and adds `,` separated items to class list in HTML. `rust foo` actually puts the text `foo` *into* the code block. `<code class="rust">foo</code>`.

All four treat characters other than `,`, ` `, `\t` the same as alphanumeric, i.e. the don't split, and are included in the parsed language string.

It seems like there's little agreement 😅 If there are other systems, that I should take a look at, let me know.

Do I need to do anything to trigger a crater run?

**Comment by jyn514** [comment]
I'm not sure if crater would tell us anything meaningful here, actually. Rustdoc doesn't warn or error when it runs into a language string it doesn't understand, so even if there were a change in behavior we wouldn't see it on crater.

**Comment by casey** [comment]
We could see test failures if attributes stop being recognized. For example, this would currently get the `should_panic` attribute, but after the change wouldn't, and would fail:

````rust
```rust&should_panic
panic!();
```
````

It would be ideal to see if the number of tests that ran changed. If they stayed the same and nothing failed, we could be confident that there were no breakages.

**Comment by Manishearth** [comment]
@jyn514 correct, but we can push a change that makes this a hard error and then crater it

**Comment by jyn514** [comment]
Got it, thanks.

@casey can you change the PR to give an error/give an assertion failure if the new and old parsing don't match? You'll have to temporarily disable the tests. Then I can start a crater run to see how often this happens in practice.

**Comment by casey** [comment]
I think this is ready for a crater run. I modified it to parse lang strings twice, once with the new splitting scheme, and once with the old, and panic if the parsed `LangString` objects don't match.

File: src/test/rustdoc/issue-32556.rs
Comment (jyn514) [inline src/test/rustdoc/issue-32556.rs:2]: I don't think you meant to change this.
```suggestion
/// ```ignore (testing rustdoc's handling of ignore)
```

File: compiler/rustc_error_codes/src/error_codes/E0761.md
Comment (jyn514) [inline compiler/rustc_error_codes/src/error_codes/E0761.md:5]:
```suggestion
```ignore (Multiple source files are required for compile_fail.)
```

**Review by jyn514 — COMMENTED** [review]
I'm building it locally to make sure it gives an assertion failure. I have a few minor nits in the meantime.

File: src/librustdoc/html/markdown.rs
Comment (jyn514) [inline src/librustdoc/html/markdown.rs:833]: This prints the assertion failure twice:

```
thread 'rustc' panicked at 'assertion failed: `(left == right)`
  left: `LangString { original: "{.no_run .example}", should_panic: false, no_run: false, ignore: None, rust: false, test_harness: false, compile_fail: false, error_codes: [], allow_fail: false, edition: None }`,
 right: `LangString { original: "{.no_run .example}", should_panic: false, no_run: true, ignore: None, rust: true, test_harness: false, compile_fail: false, error_codes: [], allow_fail: false, edition: None }`: New token split differs from old split.
New: LangString { original: "{.no_run .example}", should_panic: false, no_run: false, ignore: None, rust: false, test_harness: false, compile_fail: false, error_codes: [], allow_fail: false, edition: None }
Old: LangString { original: "{.no_run .example}", should_panic: false, no_run: true, ignore: None, rust: true, test_harness: false, compile_fail: false, error_codes: [], allow_fail: false, edition: None }', src/librustdoc/html/markdown.rs:829:9
```

File: src/librustdoc/html/markdown/tests.rs
Comment (casey) [reply src/librustdoc/html/markdown/tests.rs:103]: They're very slightly different, the latter two have a trailing comma, and a trailing comma and space.

File: src/librustdoc/html/markdown/tests.rs
Comment (casey) [reply src/librustdoc/html/markdown/tests.rs:103]: Good catch, I'll add them back with the correct syntax.

File: src/librustdoc/html/markdown.rs
Comment (casey) [reply src/librustdoc/html/markdown.rs:833]: Fixed

File: src/librustdoc/html/markdown/tests.rs
Comment (casey) [reply src/librustdoc/html/markdown/tests.rs:103]: Fixed

File: src/librustdoc/html/markdown.rs
Comment (casey) [reply src/librustdoc/html/markdown.rs]: Fixed

**Comment by casey** [comment]
Those two comment changes you noticed were actually necessary to get the parser to return the same result for those two lines. Using the old splitter `foo)` would split as `foo`, but the new one splits it as `foo)`. This led the new splitter to ignore those tokens as invalid, which changed the way the lang string was parsed. I think the difference in parsing actually doesn't make a difference, it only changes whether or not the `rust` bool is set on the lang string, which doesn't make a difference if it's already `ignore`ed.

Other than that, hopefully I addressed all your comments.

**Comment by nikomatsakis** [comment]
@rfcbot reviewed 

I would be happy to match github -- it sounds like the newly proposed behavior is "closer" to what other markdown processors do, in any case.

**Comment by jyn514** [comment]
@craterbot run mode=rustdoc

**Comment by jyn514** [comment]
```
     Checking rustc_plugin_impl v0.0.0 (/checkout/compiler/rustc_plugin_impl)
 Documenting rustc_plugin_impl v0.0.0 (/checkout/compiler/rustc_plugin_impl)
thread 'rustc' panicked at 'assertion failed: `(left == right)`
  left: `LangString { original: "ignore(cross-crate-imports)", should_panic: false, no_run: false, ignore: None, rust: false, test_harness: false, compile_fail: false, error_codes: [], allow_fail: false, edition: None }`,
 right: `LangString { original: "ignore(cross-crate-imports)", should_panic: false, no_run: false, ignore: All, rust: true, test_harness: false, compile_fail: false, error_codes: [], allow_fail: false, edition: None }`: New token split differs from old token split', src/librustdoc/html/markdown.rs:829:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```
Hmm, maybe rustdoc should also split on parentheses? I'd also be ok with just changing the compiler ignores and leaving rustdoc as-is, I expect we use `// ignore` more than most of the ecosystem.

**Comment by casey** [comment]
> Hmm, maybe rustdoc should also split on parentheses?

I could also make it so that anything inside of matched parentheses is ignored, so it really functions as a comment, which seems to be the convention used in the rust codebase.

This is a little weird, and precludes using parentheses in attributes. The latter is unlikely, but I can imagine it happening if there is some kind of attribute that has a free text in it, e.g. `rust label=foo(bar)`, or if for some reason we introduced an attribute with a function-call-like syntax, e.g. `rust foo(1)`.

> I'd also be ok with just changing the compiler ignores and leaving rustdoc as-is, I expect we use // ignore more than most of the ecosystem.

I grepped, and it looks like the one that failed is the only such line that's likely to fail, and could be fixed by adding a space after ignore.

**Comment by jyn514** [comment]
@casey let's add a space for now then, and we can change the parsing if it turns out in the crater run that other people are using this.

**Comment by casey** [comment]
@jyn514 Sounds good, just added spaces to the two failing langstrings.

**Comment by jyn514** [comment]
@craterbot run mode=rustdoc

:crossed_fingers:

**Comment by jyn514** [comment]
Looks like most of the regressions come from parsing `.ignore` differently from `ignore` now.

**Comment by casey** [comment]
This is my first time reading a crater report. Does it deduplicate failures in dependencies? I.e. if two root crates fail because of the same dependency, does it report two failures or one?

I downloaded and grepped through the regressed crate build failures, and there are 1261 failures, and all but 176 are for the string `.ignore`.

There are a small number of unique cases. Here they are, along with the number of times they appear:

```
1085 .ignore
  94 {rust}
  20 {rust,ignore}
  14 {.should_panic}
   9 {.rust}
   5 rust{compile_fail}
   5 .rust
   4 rust#ignore
   3 [no_run]
   3 !ignore
   2 rust.ignore
   2 ignore{.should_panic}
   2 # compile_fail
   1 ¨
   1 {.ignore .rust}
   1 should_panic(expected = \"Method sequence violation\")
   1 rust{.should_panic}
   1 rust{.ignore}
   1 rust{.example}
   1 rust\u{328}
   1 rust,should_panic=\"Assert CLI failure\"
   1 rust(ignore)
   1 no_run(
   1 ignore=
   1 ;
   1 //!
```

I have no idea how to put these numbers in perspective, so very curious what others make of this.

File: compiler/rustc_error_codes/src/error_codes/E0761.md
Comment (jyn514) [inline compiler/rustc_error_codes/src/error_codes/E0761.md:5]: This can be reverted, I think.

File: src/librustdoc/html/markdown.rs
Comment (jyn514) [inline src/librustdoc/html/markdown.rs:834]: I think most of this can been reverted now that we've done a crater run.

File: src/test/rustdoc/issue-32556.rs
Comment (jyn514) [inline src/test/rustdoc/issue-32556.rs:2]: This can be reverted.

**Review by jyn514 — COMMENTED** [review]
I think this is still waiting on some more people from lang team/rustdoc to sign off. Is anyone concerned by the regressions in crater or do they all look niche enough not to worry about? We could fix almost all of them by also splitting on `{}.`, I'd be fine with that.

File: compiler/rustc_error_codes/src/error_codes/E0761.md
Comment (casey) [reply compiler/rustc_error_codes/src/error_codes/E0761.md]: Reverted.

File: src/librustdoc/html/markdown.rs
Comment (casey) [reply src/librustdoc/html/markdown.rs]: Done!

File: src/test/rustdoc/issue-32556.rs
Comment (casey) [reply src/test/rustdoc/issue-32556.rs]: Reverted.

**Comment by jyn514** [comment]
ping @GuillaumeGomez @ollie27 @pnkfelix @withoutboats - this is waiting on your review: https://github.com/rust-lang/rust/pull/78429#issuecomment-716942661

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
Sorry, completely forgot about it. Code looks good to me!

**Comment by casey** [comment]
I can't figure out how to run the tests in `src/librustdoc/html/markdown/tests.rs`.

I thought that `x.py src/test/rustdoc` would run them, but I inserted intentional errors into `src/librustdoc/html/markdown/tests.rs`, and it looks like it doesn't.

`cargo test -p rustdoc` and `cargo test` while in `src/librustdoc` don't work either:

```
error[E0463]: can't find crate for `lazy_static`
  --> src/librustdoc/lib.rs:23:1
   |
23 | extern crate lazy_static;
   | ^^^^^^^^^^^^^^^^^^^^^^^^^ can't find crate

error: aborting due to previous error
```

Is there another way to run these tests?

**Comment by jyn514** [comment]
@casey try `x.py test --stage 0 src/librustdoc`.

**Comment by casey** [comment]
@jyn514 Ahhh, thank you! Tests should be fixed.

**Comment by jyn514** [comment]
ping @ollie27 @pnkfelix @withoutboats - this is waiting on your review: https://github.com/rust-lang/rust/pull/78429#issuecomment-716942661. The crater run is here if you want to read it: https://github.com/rust-lang/rust/pull/78429#issuecomment-726485338

**Comment by ollie27** [comment]
Sorry for the slow reply.

Given the amount of breakage found by crater I do think we should continue to support the `{.should_panic}` like syntax by splitting on `{`, `}` and `.` at least. Although, does anyone know where that syntax originally came from?

**Comment by jyn514** [comment]
> Although, does anyone know where that syntax originally came from?

It comes from pandoc: https://pandoc.org/MANUAL.html#extension-fenced_code_attributes. The first commit adding it was fefdb63c4c2b078c43836e2ae9d7ffcaeec32890.

@casey can you change this to split on `{}.` as well? Thank you for sticking with this :heart:

**Comment by casey** [comment]
> It comes from pandoc: https://pandoc.org/MANUAL.html#extension-fenced_code_attributes. The first commit adding it was [fefdb63](https://github.com/rust-lang/rust/commit/fefdb63c4c2b078c43836e2ae9d7ffcaeec32890).

Nice sleuthing!

> @casey can you change this to split on `{}.` as well?

I can definitely do that, although what do you think about doing the minimum possible to confirm to the pandoc syntax.

I could:

1. Remove `{` and `}` if they are the first and last character, respectively, of the langstring.
2. Remove a single leading `.` if present, from each token.

This would allow `{`, and `}` to be used freely within lang strings. (Assuming they don't also start with a `{`, in which case a trailing `}` would also be stripped.)

And `.` could also be used within lang string tokens and values, so `foo.bar=baz` or `baz=foo.bar` would both be possible, if so desired, in future extensions.

> Thank you for sticking with this ❤️

You bet! Thank you for your patience and review.

**Comment by casey** [comment]
Additionally, I could also add warnings when the special behavior of `{}.` was triggered, so that it could be removed some time in the future.

**Comment by jyn514** [comment]
Personally, I don't think there's a need to remove `{}.`, there's no proposed use for them and I'd prefer not to have things change meaning.

I don't have a preference for whether to always split on `{}.` or the most target approach you suggested. I slightly favor the more target approach just because it gives rustdoc more flexibility.

**Comment by casey** [comment]
> Personally, I don't think there's a need to remove `{}.`, there's no proposed use for them and I'd prefer not to have things change meaning.

I definitely wouldn't propose changing the meaning, but one benefit of just adding a warning would be discouraging people from inadvertently relying on the special behavior of `{}.`. It might be annoying if a the existing code started throwing warnings though, so I think whether or not a warning is a good idea would also strongly depend on whether or not the existing uses would get fixed eventually, or would just throw warnings forever.

> I don't have a preference for whether to always split on `{}.` or the most targeted approach you suggested. I slightly favor the more targeted approach just because it gives rustdoc more flexibility to add things later.

Sounds good, I'll update this PR with the targeted approach.

**Comment by casey** [comment]
@jyn514 I just pushed a change that strips a surrounding `{}` from the langstring, and a leading `.` from each token, along with some test cases.

**Comment by jyn514** [comment]
Thanks, looks great! @ollie27 can you take another look? I think we'll be ready to go after that :)

@casey in the meantime do you mind squashing the commits to something fewer than 24?

**Comment by casey** [comment]
> @casey in the meantime do you mind squashing the commits to something fewer than 24?

Yeah, 24 is a bit excessive for this change 😅 I squashed to one commit, and improved the commit message.

**Comment by dtolnay** [comment]
I've approved on behalf of withoutboats, as per https://github.com/rust-lang/team/pull/526.

**Comment by casey** [comment]
Awesome! Thank you @jyn514 for helping me get this through!

> As the automated representative of the governance process, I would like to thank the author for their work and everyone else who contributed.

I love how polite rfcbot is 😊

**Comment by jyn514** [comment]
You're very welcome! Thanks for sticking with it so long :)

**Pull request merged**
